kong/tools/utils.lua

```diff
@@ -1265,24 +1264,31 @@ do
   -- so use 64KB - 1 instead
   local GZIP_CHUNK_SIZE = 65535
 
-  local function gzip_helper(op, input)
-    local f = stringio_open(input)
-    local output_table = {}
-    local output_table_n = 0
+  local input_buffer = buffer.new()
+  local output_buffer = buffer.new()
 
-    local res, err = op(function(size)
-      return f:read(size)
-    end,
-    function(res)
-      output_table_n = output_table_n + 1
-      output_table[output_table_n] = res
-    end, GZIP_CHUNK_SIZE)
+  local function prepare_buffers(input)
+    input_buffer:set(input)
+    output_buffer:reset()
+  end
+
+  local function read_input_buffer(size)
+    local data = input_buffer:get(size)
+    return data ~= "" and data or nil
+  end
 
-    if not res then
+  local function write_output_buffer(data)
+    return output_buffer:put(data)
+  end
+
+  local function gzip_helper(inflate_or_deflate, input)
+    prepare_buffers(input)
+    local ok, err = inflate_or_deflate(read_input_buffer, write_output_buffer, GZIP_CHUNK_SIZE)
+    if not ok then
       return nil, err
     end
 
-    return concat(output_table)
+    return output_buffer:get()
```

Comment:
TODO: do we want to free the buffers before returning?

Reply:
Given that the buffers are module global, freeing them would make some sense.  I find it slightly unhygienic that the buffers are reused across calls.  Do you have evidence that `prepare_buffers()` is actually faster than creating new buffers for every invocation of `gzip_helper`?  If they were not reused, they'd be quickly collected automatically.

Reply:
@hanshuebner,

So options are:
- `:free` them
- always create them
- keep the memory they allocated and reuse the memory whenever new need comes (current implementation), e.g. on DP that receives config from CP usually needs similar buffer -> less garbage, less memory allocations. 
- allow buffers to be given to functions

Reply:
Probably the `2.` is the way to go.

Reply:
`2.` is simple, but `3.` may be faster - The memory reuse is a good point, however long-lived allocations also have the risk of adding to heap fragmentation.

Reply:
Yes, and not only that, long-lived allocations have also a problem of persistent memory usage. That was my original concern that does 100 MB config also cause 100 MB persistent memory usage because of these buffers.

Reply:
OK, so then I'd say go with option `2.` for now under the assumption that allocating buffers is fast and freeing them quickly is better than trying to reuse them.